lib/IRGen/GenOpaque.cpp

```diff
@@ -629,6 +629,82 @@ void IRGenFunction::emitDeallocateDynamicAlloca(StackAddress address) {
   Builder.CreateCall(stackRestoreFn, savedSP);
 }
 
+llvm::Value *IRGenFunction::emitStackAlloc(llvm::Value *size,
+                                           llvm::Value *align) {
+  auto returnType = IGM.Int8Ty;
+  auto returnPtrType = returnType->getPointerTo();
+  StackAddress stackPointer;
+  auto nullPointer = llvm::Constant::getNullValue(returnPtrType);
+
+  // Check that this function type can emit a stack allocation safely.
+  if (isAsync() || isCoroutine()) {
+    // These types of functions don't actually stack-allocate, and support for
+    // them is not needed at this time.
+    return nullPointer;
+  }
+
+  // Check that the caller-supplied alignment is usable.
+  Alignment resolvedAlign;
+  if (auto alignConst = dyn_cast<llvm::ConstantInt>(align)) {
+    // Ensure the supplied alignment is itself a power of 2.
+    auto alignConstValue = alignConst->getSExtValue();
+    if (!llvm::isPowerOf2_64(alignConstValue)) {
+      alignConstValue = llvm::NextPowerOf2(alignConstValue);
+    }
+
+    if (alignConstValue > 0) {
+      resolvedAlign = Alignment(alignConstValue);
+
+    } else {
+      // The specified alignment is nonsensical (negative, zero, or too big.)
+      return nullPointer;
+    }
+
+  } else {
+    // The caller specified a non-constant alignment value.
+    return nullPointer;
+  }
+
+  // Check that the requested size is small enough to fit on the stack. If size
+  // is constant, the compiler can optimize this away. If it is not constant,
+  // the check must be performed at runtime. Equivalent to:
+  //
+  // void *ptr;
+  // if (size <= stackLimit) {
+  //   ptr = alloca(...);
+  // } else {
+  //   ptr = nullptr;
+  // }
+  auto stackLimit = llvm::ConstantInt::getSigned(IGM.SizeTy,
+                                                 IGM.IRGen.Opts.StackAllocSizeLimit);
+  auto predSmallEnough = Builder.CreateICmpSLE(size, stackLimit);
+  auto smallEnoughBB = createBasicBlock("stack_alloc_ok");
+  auto tooBigBB = createBasicBlock("stack_alloc_too_big");
+  auto doneBB = createBasicBlock("stack_alloc_done");
+
+  Builder.CreateCondBr(predSmallEnough, smallEnoughBB, tooBigBB); {
+    Builder.emitBlock(smallEnoughBB); {
+      stackPointer = emitDynamicAlloca(returnType, size, resolvedAlign,
+                                       "stack_alloc");
+    } Builder.CreateBr(doneBB);
+
+    Builder.emitBlock(tooBigBB); {
+      // Here's where we might call out to the stdlib to ask for assistance.
+      // The stdlib could e.g. check thread conditions to determine if it's safe
+      // to issue an alloca, at which point we'd emit one.
+    } Builder.CreateBr(doneBB);
+  } Builder.emitBlock(doneBB);
+
+  auto result = Builder.CreatePHI(returnPtrType, 2);
+  result->addIncoming(stackPointer.getAddressPointer(), smallEnoughBB);
+  result->addIncoming(nullPointer, tooBigBB);
+
+  // FIXME: Box in an Optional. CreateBitCast?
+  // result = boxSomehow(result);
```

Comment:
For an optional, looks like `%i = ptrtoint i8* %p to i64` should be the right thing on 64-bit. I don't know if there is a better function here or if you should directly have a bit-cast...

Reply:
LLVM vomits if I try to emit a `ptrtoint` here, so there's something extra happening that I don't know how to resolve.

Reply:
The return type would also need to be changed; right now it's `i8 *`. Maybe that's the source of the issue?

Reply:
I'm honestly not certain—I need somebody better-versed in this stuff than me to chime in. Some comments elsewhere in the stdlib make me think this is an unavoidable problem. For now, the hackery I'm doing is good enough for testing, though.

Reply:
We discussed off of GitHub. No obvious solution is presenting itself at this time.